sale_order_import_edifact/wizard/sale_order_import.py

```diff
@@ -0,0 +1,248 @@
+import json
+import logging
+import os
+from base64 import b64decode
+from collections import defaultdict
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+logger = logging.getLogger(__name__)
+
+
+class SaleOrderImport(models.TransientModel):
+    _inherit = "sale.order.import"
+
+    import_type = fields.Selection(
+        selection_add=[("edifact", "EDIFACT")], ondelete={"edifact": "cascade"}
+    )
+    state = fields.Selection(
+        selection_add=[("tech", "Technical")], ondelete={"tech": "set default"}
+    )
+    parsed_json = fields.Text()
+    edifact_ok = fields.Boolean(compute="_compute_edifact_ok", store=True)
+
+    @api.depends("order_file", "order_filename")
+    def _compute_edifact_ok(self):
+        for rec in self:
+            if rec.order_file and rec.order_filename:
+                conf_ext = self.env.context.get("file_ext", "txt,d96a")
+                extensions = conf_ext.split(",")
+                path, ext = os.path.splitext(self.order_filename)
+                ok = ext and ext[1:] in extensions
+                if not ok:
+                    ok = b64decode(self.order_file[:4]) == b"UNB"
+                rec.edifact_ok = ok
+            else:
+                rec.edifact_ok = False
+
+    @api.onchange("order_file")
+    def order_file_change(self):
+        # Support get context if using EDI machinery
+        res = super().order_file_change()
+        self.price_source = self.env.context.get("price_source")
+        return res
+
+    def parse_order_button(self):
+        self.ensure_one()
+        edifact_model = self.env["base.edifact"]
+        order_file_decoded = b64decode(self.order_file)
+        parsed_obj = dict(edifact=None, order=None)
+        if self.edifact_ok:
+            parsed_obj["edifact"] = edifact_model.pydifact_obj(order_file_decoded)
+
+        parsed_order = self.parse_order(
+            order_file_decoded, self.order_filename, self.partner_id
+        )
+        del parsed_order["attachments"]
+        parsed_obj["order"] = parsed_order
+        self.write(
+            {
+                "parsed_json": json.dumps(parsed_obj, indent=4, default=str),
+                "state": "tech",
+            }
+        )
+        action = self.env["ir.actions.act_window"]._for_xml_id(
+            "sale_order_import.sale_order_import_action"
+        )
+        action["res_id"] = self.id
+        return action
+
+    def _get_supported_types(self):
+        # Add more types for EDIFACT
+        res = super()._get_supported_types()
+        res.update({"edifact": ("text/plain")})
+        return res
+
+    @api.model
+    def parse_edifact_order(self, filecontent, detect_doc_type=False):
+        # https://github.com/nerdocs/pydifact/blob/master/pydifact/segmentcollection.py
+        if detect_doc_type:
+            if not self.edifact_ok:
+                return None
+            return self.env.context.get("doc_type", "rfq")
```

Comment:
I'm not sure to get this...

Reply:
Because, to my knowledge, this editfact standard does not contain `rfq` or `order` information. So we set the default as `rfq`

If this module integrates with EDI, we will get the value of `doc_type` context from EDI machinery. Also, if in EDI machinery, we set  `'doc_type: 'order'`, maybe we need to set `default_price_source` because of [this](https://github.com/OCA/edi/blob/32a0e4938f755cbcce815c5ab613794056a8ad2c/sale_order_import/wizard/sale_order_import_view.xml#L62)

cc @rmorant , could you help me to answer:
- Why do you set `doc_type = 'rfq'` and `price_source = 'order'` fixedly in v14? 

(Notes: In v16, `sale_order_import_edifact` disconnects from the EDI framework)

Reply:
pls see my comment above regarding defaults via ctx